### PR TITLE
Add a proposal about inference of the modifier `required`

### PR DESCRIPTION
**Pull request by eernstg**
This PR adds a feature specification of the feature whereby `required` is added implicitly to parameters of concrete function and method declarations when it would be an error to not have that modifier.

### CODE REVIEW

**Comment by eernstg** [comment]
@dart-lang/language-team, WDYT? Here is a proposal for inference of requiredness. It is just a consolidation of a rather large number of discussions and proposals in this area. The specification itself is just a handful of lines. ;-)

**Review by jakemac53 — APPROVED** [review]

**Review by leafpetersen — APPROVED** [review]

**Review by kallentu — APPROVED** [review]

File: working/0015-infer-required/feature-specification.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  String address,
+  String phoneNumber,
+  String name,
+  String? avatarUrl,
+});
+```
+
+It seems likely that it will be true quite often that the required, named
+parameters are exactly the ones whose type is potentially non-nullable,
+which means that the migration will simply be to delete every occurrence of
```
Comment (natebosch) [inline working/0015-infer-required/feature-specification.md:140]: If we have an automated migration I think we should make give it full fidelity - a `required` which would not be inferred shouldn't be removed.
I don't know of specific examples, but I think it's reasonable for existing code to use `required` to force callers to explicitly pass `null` when that is the intention.

**Review by natebosch — APPROVED** [review]

File: working/0015-infer-required/feature-specification.md (hunk 1 above)
Comment (eernstg) [reply working/0015-infer-required/feature-specification.md:140]: Right, a mechanical migration should certainly be semantics preserving. Added a sentence to clarify this point.

**Comment by eernstg** [comment]
Thanks! I'll land this document now, and we would then use additional PRs in order to make changes as needed.

**Pull request merged**